### PR TITLE
[NEURON][PT][BUILD] Add neuron sdk2.4.0

### PR DESCRIPTION
**Pull request by aws-vrnatham**
Signed-off-by: Venky Natham <vrnatham@amazon.com>

*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Description
Add the new neuron sdk release and also SM tests on multiple nodes

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pytorch/training/docker/1.11/py3/sdk2.4.0/Dockerfile.neuron.os_scan_allowlist.json [hunk 1]
```diff
@@ -0,0 +1,708 @@
+{
```
Comment (tejaschumbalkar) [inline pytorch/training/docker/1.11/py3/sdk2.4.0/Dockerfile.neuron.os_scan_allowlist.json:1]: Can we review the packages in this file and remove if its been fixed by new sdk release, or check the job logs for the latest scan report

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_neuron.py [hunk 2]
```diff
@@ -17,23 +17,90 @@
...
 
-RESOURCE_PATH = os.path.join(os.path.dirname(__file__), '..', '..', 'resources')
+def retry_if_value_error(exception):
+    """Return True if we should retry (in this case when it's an ValueError), False otherwise"""
+    return isinstance(exception, ValueError)
 
+@retry(
+    stop_max_attempt_number=360, wait_fixed=10000, retry_on_exception=retry_if_value_error,
+)
+def invoke_neuron_helper_function(ecr_image, sagemaker_regions, helper_function, helper_function_args):
```
Comment (tejaschumbalkar) [inline test/sagemaker_tests/pytorch/training/integration/sagemaker/test_neuron.py:31]: If we are using just one region, we do not need this function

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_neuron.py [hunk 3]
```diff
@@ -17,23 +17,90 @@
...
+@pytest.mark.processor("neuron")
+@pytest.mark.model("unknown_model")
+@pytest.mark.neuron_test
+def test_neuron_allreduce_distributed(framework_version, ecr_image, sagemaker_regions, neuron_efa_instance_type):
+    function_args = {
+            'framework_version': framework_version,
+            'instance_type': neuron_efa_instance_type,
+            'instance_count': 2,
+        }
+    invoke_neuron_helper_function(ecr_image, sagemaker_regions, _test_neuron_allreduce_distributed, function_args)
```
Comment (tejaschumbalkar) [inline test/sagemaker_tests/pytorch/training/integration/sagemaker/test_neuron.py:69]: Same, as said above, we do not need the helper function, add the contents of `_test_neuron_allreduce_distributed` function here directly

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_neuron.py [hunk 4]
```diff
@@ -17,23 +17,90 @@
...
+@pytest.mark.processor("neuron")
+@pytest.mark.model("mlp")
+@pytest.mark.neuron_test
+def test_neuron_mlp_distributed(framework_version, ecr_image, sagemaker_regions, neuron_efa_instance_type):
+    function_args = {
+            'framework_version': framework_version,
+            'instance_type': neuron_efa_instance_type,
+            'instance_count': 2,
+        }
+    invoke_neuron_helper_function(ecr_image, sagemaker_regions, _test_neuron_mlp_distributed, function_args)
```
Comment (tejaschumbalkar) [inline test/sagemaker_tests/pytorch/training/integration/sagemaker/test_neuron.py:80]: Applicable for all tests

File: test/dlc_tests/ecs/pytorch/training/test_ecs_pytorch_training.py [hunk 5]
```diff
@@ -13,7 +13,9 @@
 PT_MNIST_TRAINING_SCRIPT = os.path.join(CONTAINER_TESTS_PREFIX, "pytorch_tests", "testPyTorch")
 PT_DGL_TRAINING_SCRIPT = os.path.join(CONTAINER_TESTS_PREFIX, "dgl_tests", "testPyTorchDGL")
 PT_S3_PLUGIN_CMD = os.path.join(CONTAINER_TESTS_PREFIX, "pytorch_tests", "testPyTorchS3Plugin")
-PT_MLP_NEURON_TRAINING_SCRIPT = os.path.join(CONTAINER_TESTS_PREFIX, "pytorch_tests", "testNeuronMlp")
+# For neuron the training script is called using torchrun. Since the trainind cmd is built based on 
+# the training script, adding torchrun here.
+PT_MLP_NEURON_TRAINING_SCRIPT = f'torchrun --nproc_per_node=2 --nnodes=1' + os.path.join(CONTAINER_TESTS_PREFIX, "pytorch_tests", "testNeuronMlp")
```
Comment (tejaschumbalkar) [inline test/dlc_tests/ecs/pytorch/training/test_ecs_pytorch_training.py:18]: is empty space needed after `--nnodes=1`?

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_neuron.py [hunk 6]
```diff
@@ -60,21 +62,21 @@ def invoke_neuron_helper_function(ecr_image, sagemaker_regions, helper_function,
 @pytest.mark.processor("neuron")
 @pytest.mark.model("unknown_model")
 @pytest.mark.neuron_test
-def test_neuron_allreduce_distributed(framework_version, ecr_image, sagemaker_regions, neuron_efa_instance_type):
+def test_neuron_allreduce_distributed(framework_version, ecr_image, sagemaker_regions):
     function_args = {
             'framework_version': framework_version,
-            'instance_type': neuron_efa_instance_type,
+            'instance_type': 'ml.trn1.32xlarge',
```
Comment (tejaschumbalkar) [inline test/sagemaker_tests/pytorch/training/integration/sagemaker/test_neuron.py:68]: instance type can be declared as a fixture like [this](https://github.com/aws/deep-learning-containers/blob/master/test/sagemaker_tests/pytorch/training/integration/sagemaker/test_smdataparallel.py#L56)

**Review by tejaschumbalkar — APPROVED** [review]
